test/Dialect/Comb/canonicalization.mlir

```diff
@@ -67,6 +67,58 @@ hw.module @andDedupLong(%arg0: i7, %arg1: i7, %arg2: i7) -> (i7) {
   hw.output %0 : i7
 }
 
+// CHECK-LABEL: hw.module @orExclusiveConcats
+hw.module @orExclusiveConcats(%arg0: i6, %arg1: i2) -> (%o: i9) {
+  // CHECK-NEXT:    %false = hw.constant false
+  // CHECK-NEXT:    %0 = comb.concat %arg1, %false, %arg0 : (i2, i1, i6) -> i9
+  // CHECK-NEXT:    hw.output %0 : i9
+  %c0 = hw.constant 0 : i3
+  %0 = comb.concat %c0, %arg0 : (i3, i6) -> i9
+  %c1 = hw.constant 0 : i7
+  %1 = comb.concat %arg1, %c1 : (i2, i7) -> i9
+  %2 = comb.or %0, %1 : i9
+  hw.output %2 : i9
+}
+
+// When two concats are or'd together and have mutually-exclusive fields, they
+// can be merged together into a single concat.
+// concat0: 0aaa aaa0 0000 0bb0
+// concat1: 0000 0000 ccdd d000
+// merged:  0aaa aaa0 ccdd dbb0
+// CHECK-LABEL: hw.module @orExclusiveConcats2
+hw.module @orExclusiveConcats2(%arg0: i6, %arg1: i2, %arg2: i2, %arg3: i3) -> (%o: i16) {
+  // CHECK-NEXT:    %false = hw.constant false
+  // CHECK-NEXT:    %0 = comb.concat %false, %arg0, %false, %arg2, %arg3, %arg1, %false : (i1, i6, i1, i2, i3, i2, i1) -> i16
+  // CHECK-NEXT:    hw.output %0 : i16
+  %c0 = hw.constant 0 : i1
+  %c1 = hw.constant 0 : i6
+  %c2 = hw.constant 0 : i1
+  %0 = comb.concat %c0, %arg0, %c1, %arg1, %c2: (i1, i6, i6, i2, i1) -> i16
+  %c3 = hw.constant 0 : i8
+  %c4 = hw.constant 0 : i3
+  %1 = comb.concat %c3, %arg2, %arg3, %c4 : (i8, i2, i3, i3) -> i16
+  %2 = comb.or %0, %1 : i16
+  hw.output %2 : i16
+}
+
+// When two concats are or'd together and have mutually-exclusive fields, they
```

Comment:
Can you add a case for an or > 2 operands?

Reply:
Yep, no problem just added a test case with 3 operands.

I also took the opportunity to rebase this on top of the latest `main`, since several bumps to llvm submodule have happened in the meantime.

Reply:
Rebasing isn't necessary if you don't have conflicts.  Squash merge is used exclusively so there is a linear history with 1 PR == 1 commit.

Reply:
Got it. I'm never sure what GitHub squash merges will do with merges in the feature branch, so I've always erred on the side of rebasing my branches when keeping up to date with the main branch, but good to know.